strawberryfields/engine.py

```diff
@@ -337,33 +337,51 @@ def _run_program(self, prog, **kwargs):
-            shots (int): number of times the program measurement evaluation is repeated
-            compile_options (Dict[str, Any]): keyword arguments for :meth:`.Program.compile`
-            modes (None, Sequence[int]): Modes to be returned in the ``Result.state`` :class:`.BaseState` object.
-                An empty sequence [] means no state object is returned. None returns all the modes.
-            state_options (Dict[str, Any]): keyword arguments for :meth:`.BaseBackend.state`
-
-        The ``kwargs`` keyword arguments are passed to :meth:`_run_program`.
+            compile_options (None, Dict[str, Any]): keyword arguments for :meth:`.Program.compile`
+            run_options (None, Dict[str, Any]): keyword arguments that are needed by the backend during execution
+                e.g., in :meth:`Operation.apply` or :meth:`.BaseBackend.state`
 
         Returns:
             Result: results of the computation
-        """
 
-        # session or feed_dict are needed by TF backend during simulation if program contains measurements
-        kwargs.update(state_options)
-        result = super()._run(program, shots=shots, compile_options=compile_options, **kwargs)
-        if isinstance(modes, Sequence) and not modes:
-            # empty sequence
-            pass
-        else:
-            result._state = self.backend.state(modes, **state_options)  # tfbackend.state can use kwargs
+        ``run_options`` can contain the following:
+
+        Keyword Args:
+            shots (int): number of times the program measurement evaluation is repeated
+            modes (None, Sequence[int]): Modes to be returned in the ``Result.state`` :class:`.BaseState` object.
+                ``None`` returns all the modes (default).
+            return_state (bool): Whether to return a :class:`.BaseState` object object within ``Result.state``.
+                Default is True.
+            eval (bool): If False, the backend returns unevaluated tf.Tensors instead of
+                numbers/arrays for returned measurement results and state (default: True). TF backend only.
+            session (tf.Session): TensorFlow session, used when evaluating returned measurement results and state.
+                TF backend only.
+            feed_dict (dict[str, Any]): TensorFlow feed dictionary, used when evaluating returned measurement results
+                and state. TF backend only.
+        """
+        compile_options = compile_options or {}
+        run_options = run_options or {}
+        run_options.setdefault("shots", 1)
+        run_options.setdefault('modes', None)
+
+        # avoid unexpected keys being sent to Operations
+        eng_run_keys = ["eval", "session", "feed_dict", "shots"]
+        eng_run_options = {key: run_options[key] for key in run_options.keys() & eng_run_keys}
+
+        result = super()._run(program, compile_options=compile_options, **eng_run_options)
```

Comment:
This might need a future rethink --- this will preclude future backends from allowing their own run options, without also needing to modify the frontend to whitelist it here.

Ideally, the idea behind a run_options dict rather than frontend defined kwargs is to support the behavior where run options can be defined by the backend.

However, since at the moment `_run` accepts `**kwargs`, this would be a significant change/refactor.